File: 2024/March/3-15 238 Product of Array Except Self/Solution1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];

        int product = 1;
        for(int i = 0; i < nums.length; i++) {
            prefix[i] = product;
            product *= nums[i];
        }

        product = 1;
        for(int i = nums.length - 1; i >= 0; i--) {
            suffix[i] = product;
            product *= nums[i];
        }

        int[] result = new int[nums.length];
        for(int i = 0; i < nums.length; i++) {
            result[i] = prefix[i] * suffix[i];
        }

        return result;
    }
}

File: 2024/March/3-15 238 Product of Array Except Self/Solution1.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [0] * len(nums)
        suffix = [0] * len(nums)

        product = 1
        for i in range(len(nums)):
            prefix[i] = product
            product *= nums[i]

        product = 1
        for i in reversed(range(len(nums))):
            suffix[i] = product
            product *= nums[i]

        result = [0] * len(nums)
        for i in range(len(result)):
            result[i] = prefix[i] * suffix[i]

        return result